bin/cron/pardot_import

```diff
@@ -197,19 +188,189 @@ def main
+  # We also need to encode plus signs in email addresses, otherwise Pardot rejects them as invalid.
+  # URI.encode does not encode plus signs, as they are valid characters in the base of a URL.
+  # (Although they are NOT valid in the query string, which is where this data is going.)
+  prospects_payload_json_encoded = prospects_payload_json_encoded.gsub("+", "%2B")
+
+  "#{config[:pardot_url]}?prospects=#{prospects_payload_json_encoded}"
+end
+
+# Make an API request with Pardot authentication, including appending auth params and refreshing Pardot API key and
+# retrying if necessary
+# @param url [String] URL to post to
+# @returns [Nokogiri::XML] XML response from Pardot
+def post_with_auth_retry(url)
+  # do the post to Pardot
+  post_request_with_auth(url)
+rescue InvalidApiKeyException
+  # If we fail with an invalid key, that probably means our API key (which is good for one hour) has expired. Try one
+  # time to request a new API key and try the post again. If that fails, that is a fatal error.
+  request_pardot_api_key
+  post_request_with_auth(url)
+end
+
+# Make an API request with Pardot authentication
+# @param url [String] URL to post to
+# @returns [Nokogiri::XML] XML response from Pardot
+def post_request_with_auth(url)
+  request_pardot_api_key if $pardot_api_key.nil?
+  # add the API key and user key parameters to the URL
+  auth_url = append_auth_params_to_url(url)
+  doc = post_request(auth_url, {})
+  doc
+end
+
+# Make an API request
+# @param url [String] URL to post to - must already include Pardot auth params in query string
+# @param params [Hash] hash of POST params (may be empty hash)
+# @returns [Nokogiri::XML] XML response from Pardot
+def post_request(url, params)
+  uri = URI(url)
+
+  # make the HTTP POST
+  response = Net::HTTP.post_form(uri, params)
+
+  # do common error handling for Pardot response
+  raise "Pardot request failed with HTTP " + response.code unless PARDOT_SUCCESS_HTTP_CODES.include?(response.code)
+
+  doc = Nokogiri::XML(response.body, &:noblanks)
+  raise "Pardot response did not return parsable XML" if doc.nil?
+
+  error_details = doc.xpath('/rsp/err').text
+  raise InvalidApiKeyException if error_details.include? "Invalid API key or user key"
+
+  status = doc.xpath('/rsp/@stat').text
+  raise "Pardot response did not include status" if status.nil?
+
+  doc
+end
+
+# Append standard Pardot auth parameters (per-session API key and fixed user key) to a
+# Pardot API request
+# @param url [String] URL to post to
+# @returns [String] URL with auth parameters appended
+def append_auth_params_to_url(url)
+  "#{url}&api_key=#{$pardot_api_key}&user_key=#{CDO.pardot_user_key}"
+end
+
+# Parse a Pardot XML response and raise an exception on the first error
+# if there is one.
+# @param doc [Nokogiri::XML] XML response from Pardot
+def raise_if_response_error(doc)
+  status = doc.xpath('/rsp/@stat').text
+  if status != "ok"
+    error_text = doc.xpath('/rsp/errors/*').first.try(:text)
```

Comment:
On Asher's first point, yes, good point, I will do that.

On the broader question of what is in the response: this is defensive code against a not well documented API. The response being parsed here is sparsely documented at best. In particular, the error response is not documented at all anywhere I can find. The API method documentation is here: http://developer.pardot.com/kb/api-version-4/prospects/ . It does a decent job about describing the inputs, but nothing at all about error responses. They have a link on that page `Error Codes & Messages` which hilariously, as of this writing, leads to a 404.

There are other specific errors where you can get multiple errors (for instance, malformed emails within a batch). We handle that case specifically, iterate through the errors, and we have something interesting to do with that error where we can handle it but still continue the process. In this case, we have hit something unexpected, we are going to raise an exception and stop. It is possible that we might have gotten a response with multiple unexpected errors, none of which we know what to do with. We will raise an exception and stop. We do log the complete response so for diagnosis, if multiple errors were returned we can see that. The method is documented that is will raise an exception on the first error.